include/networkit/centrality/GedWalk.hpp

```diff
@@ -0,0 +1,231 @@
+/* GedWalk.hpp
+ *
+ *  Created on: 18.6.2018
+ *     Authors: Eugenio Angriman <angrimae@hu-berlin.de>
+ *              Alexander van der Grinten <avdgrinten@hu-berlin.de>
+ */
+
+// networkit-format
+
+#ifndef NETWORKIT_CENTRALITY_GED_WALK_HPP_
+#define NETWORKIT_CENTRALITY_GED_WALK_HPP_
+
+#include <networkit/algebraic/CSRMatrix.hpp>
+#include <networkit/base/Algorithm.hpp>
+#include <networkit/graph/Graph.hpp>
+
+#include <tlx/container/d_ary_addressable_int_heap.hpp>
+
+namespace NetworKit {
+
+class GedWalk final : public Algorithm {
+
+    using walks = double;
+    using uchar = unsigned char;
+
+public:
+    enum GreedyStrategy { lazy, stochastic };
+
+    enum BoundStrategy { no, spectral, geometric, adaptiveGeometric };
+
+    double stocEpsilon = 0.1;
+
+private:
+    const Graph *G;
+    const count k;
+    const double epsilon;
+    double alpha;
+    const BoundStrategy boundStrategy;
+    GreedyStrategy greedyStrategy;
+    double spectralDelta;
+    count degOutMax;
+    count degInMax;
+    std::vector<double> alphas;
+
+    // Size of pathsIn / pathsOut and related vectors.
+    count allocatedLevels = 15;
+
+    double sigmaMax;
+
+    // Similar to gainScore, gainBound and gainW, but for the entire group.
+    // These values are always maintained exactly.
+    double groupScore; // Score of the group.
+    walks groupW;      // Number of walks on the last level.
+    walks groupBound;  // Upper bound on the score of the group.
+
+    walks graphW;
+
+    count nLevels = 2;
+    std::vector<node> group;
+    std::vector<unsigned char> inGroup;
+    std::vector<unsigned char> isExact;
+    std::vector<unsigned char> nodesToPick;
+    std::vector<std::vector<walks>> pathsIn, pathsOut;
+    std::vector<std::vector<walks>> pathsHit, pathsMiss;
+
+    // Unless isExact[x], these values might be upper bounds on the true values.
+    std::vector<double> gainScore; // Marginal score of the group.
+    std::vector<walks> gainW;      // Number of marginal walks on the last level.
+    std::vector<double> gainBound; // Upper bound on the marginal score.
+
+    struct CompareScore {
+    public:
+        CompareScore(const std::vector<double> &score) : score(score) {}
+        bool operator()(node x, node y) const { return score[x] > score[y]; }
+
+    private:
+        const std::vector<double> &score;
+    };
+
+    struct EvaluationResult {
+        double score;
+        walks w;
+    };
+
+    tlx::d_ary_addressable_int_heap<node, 2, CompareScore> scoreQ{CompareScore(gainScore)};
+    tlx::d_ary_addressable_int_heap<node, 2, CompareScore> boundQ{CompareScore(gainBound)};
+    CSRMatrix gMat;
```

Comment:
I think it makes more sense to use a graph here. Yes, a CSR will give slightly better performance, but it requires an entire copy of the graph.

Reply:
Alright now GedWalk does not use an additional CSR matrix anymore.